src/common/src/util/value_encoding/row_encoding.rs

```diff
@@ -0,0 +1,221 @@
+use std::collections::HashMap;
+
+use super::*;
+use crate::catalog::ColumnId;
+use crate::row::Row;
+
+#[derive(Clone, Copy)]
+enum Width {
+    Mid(u8),
+    Large(u16),
+    Extra(u32),
+}
+
+struct RowEncoding {
+    flag: u8,
+    offsets: Vec<u8>,
+    buf: Vec<u8>,
+}
+
+impl RowEncoding {
+    fn new() -> Self {
+        RowEncoding {
+            flag: 0b_1000_0000,
+            offsets: vec![],
+            buf: vec![],
+        }
+    }
+
+    fn set_width(&mut self, datum_num: &Width) {
+        match datum_num {
+            Width::Mid(_) => {
+                self.flag |= 0b0100;
+            }
+            Width::Large(_) => {
+                self.flag |= 0b1000;
+            }
+            Width::Extra(_) => {
+                self.flag |= 0b1100;
+            }
+        }
+    }
+
+    fn set_big(&mut self, maybe_offset: Vec<usize>, max_offset: usize) {
+        self.offsets = vec![];
```

Comment:
```suggestion
```

Reply:
- We can calculate the `max_offset` in this function, to avoid possible invalid combinations in the interface design.
- We don't rely on the ownership of the `Vec` in this function. So although we can directly move the `Vec` in, it'd be better if we take `&[usize]`.